### PR TITLE
Upgrade folly to v2023.12.04.00 (from v2022.11.14.00)

### PR DESCRIPTION
**Pull request by assignUser**
This version no longer uses deprecated openssl functions. (the actual change is in an earlier version but I think it makes sense to  bump as far as possible while we are at it)

### CODE REVIEW

**Comment by mbasmanova** [comment]
@assignUser Jacob, thank you for the upgrade. Would you rebase and resolve conflicts?

**Comment by assignUser** [comment]
The nodiscard was just added in https://github.com/facebook/folly/commit/b7af40fe147449f08dacf28eb7468ee1e1e69c1f so let's try a bit earlier version.

**Comment by majetideepak** [comment]
Masha and I magically chose `v2023.10.30.00` and it worked. Try that if this does not work.

**Comment by assignUser** [comment]
>  magically chose

:mage: :grinning: nice. Yeah will try if this one fails

**Comment by majetideepak** [comment]
And don't forget to update the setup scripts as well.

**Comment by majetideepak** [comment]
> chose `v2023.10.30.00` and it worked

This worked only on MacOS though. Linux was not tested.

**Comment by assignUser** [comment]
hmm @majetideepak any idea what could cause the glog issue?

**Comment by majetideepak** [comment]
I found this link https://github.com/gflags/gflags/issues/203
We have to investigate further.

**Comment by majetideepak** [comment]
@assignUser I see that both gflags and glog are being built together even though there is a gflags dependency for glog.
This is likely causing the glog issue.
```
[46/2257] Building CXX object _deps/gflags-build/CMakeFiles/gflags_nothreads_static.dir/src/gflags_completions.cc.o
[47/2257] Building CXX object _deps/glog-build/CMakeFiles/glogbase.dir/src/demangle.cc.o
```

**Comment by assignUser** [comment]
Oh yeah that might be it! I'll check it out!

**Comment by assignUser** [comment]
@majetideepak looks like you were right! But we seem to have some other errors now... I had to add a new warning suppression to the expression fuzzer test (see commit message for the error) but we should probably just fix that.

**Comment by majetideepak** [comment]
I am having a hard time trying to understand these warnings.
Where does `facebook::velox::test::ExpressionFuzzer::Options::skipFunctions` come from?
```
/velox/velox/expression/tests/ExpressionFuzzerUnitTest.cpp: In lambda function:
/velox/velox/expression/tests/ExpressionFuzzerUnitTest.cpp:107:35: error: missing initializer for member 'facebook::velox::test::ExpressionFuzzer::Options::skipFunctions' [-Werror=missing-field-initializers]
  107 |         {{.maxLevelOfNesting = 5}}};
      |                                   ^
/velox/velox/expression/tests/ExpressionFuzzerUnitTest.cpp: In member function 'virtual void facebook::velox::test::ExpressionFuzzerUnitTest_exprBank_Test::TestBody()':
/velox/velox/expression/tests/ExpressionFuzzerUnitTest.cpp:133:51: error: missing initializer for member 'facebook::velox::test::ExpressionFuzzer::Options::skipFunctions' [-Werror=missing-field-initializers]
  133 |         {{.maxLevelOfNesting = maxLevelOfNesting}}};
```

**Comment by mbasmanova** [comment]
Is this relevant?

```
git show b9367aa14

Author: Laith Sakka <lsakka@meta.com>
Date:   Mon Dec 4 18:01:07 2023 -0800

    Back out "make skipFunctions, onlyFunctions, and specialForms as ExpressionFuzzer options."

    Summary:
    Dont have time to investigate will backout and now and fix later.
    Original commit changeset: c4bb6b830dc0
```

CC: @kevinwilfong @laithsakka

**Comment by majetideepak** [comment]
@mbasmanova thanks for the pointer. That is relevant. 
@assignUser rebasing with main should fix the warnings your are seeing.
Do you know how to fix the MacOS build failure related to OpenSSL?

File: velox/exec/tests/SparkAggregationFuzzerTest.cpp
Comment (assignUser) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp:48]: because `init` is what's marked as deprecated:
```cpp
[[deprecated("Use the RAII version Init")]] void init(
    int* argc, char*** argv, bool removeFlags = true);
[[deprecated("Use the RAII version Init")]] void init( 
```
https://github.com/facebook/folly/blob/main/folly/init/Init.h#L99C1-L102

If this is an issue for that specific test we could disable it just for that target with `target_compile_options( ... -Wno-deprecated-declarations)`. But I did not want to do that for everything.

**Comment by assignUser** [comment]
reverting the init change to the fuzzer files and adding `-Wno-deprecated-declarations` to that entir dir works but we should probably move those over to the new init version. I will create an issue.

**Comment by assignUser** [comment]
@majetideepak I fixed the benchmark issue but now there is this error https://github.com/facebookincubator/velox/actions/runs/7108997656/job/19353240377?pr=7700#step:14:150

Could you have a look?

**Comment by assignUser** [comment]
I had previously added some cmake to mark folly headers as system headers to suppress warnings in folly headers (e.g. the openssl deprecation warning we saw pop up in the mac build) but that did apparently not work. I have now implemented an alternative approach that should work.

**Comment by assignUser** [comment]
The cci macos build is still picking up the system openssl 3.1 even with multiple path hints set... https://app.circleci.com/pipelines/github/facebookincubator/velox/40197/workflows/f0a78360-aac9-4db8-b298-035ebfc9c122/jobs/274225?invite=true#step-108-22961_124

**Comment by majetideepak** [comment]
@assignUser Looking at the deprecated OpenSSL errors, does it mean we still require OpenSSL 1.1?

File: velox/benchmarks/basic/LikeFunctionsBenchmark.cpp
Comment (majetideepak) [inline velox/benchmarks/basic/LikeFunctionsBenchmark.cpp:236]: @assignUser the new benchmark issue we are seeing is due to this change. https://github.com/facebookincubator/velox/actions/runs/7108997656/job/19353240377?pr=7700#step:14:150

Comment (majetideepak) [reply velox/benchmarks/basic/LikeFunctionsBenchmark.cpp:236]: I wonder why the RAII version is causing failures. 
I will try to investigate if there is something fundamental lurking around.

File: velox/exec/tests/SparkAggregationFuzzerTest.cpp
Comment (majetideepak) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp]: @assignUser, `folly::Init{&argc, &argv, true};` is causing the Init object to be destroyed immediately and not at the end of main. That is why we are seeing issues with benchmark and fuzzer.
Looking at the RAII folly change here https://github.com/facebook/folly/commit/7bf1486094cccb266e789a378d8e5f91e3cb7780#diff-fe24748541a4b73975e31bb08821bb15c783ecdd74e2057d256c4cd1b893ba92
`folly::init(&argc, &argv);` is deprecated and the RAII version is `folly::Init init(&argc, &argv);`
The later does call the RAII API.

**Comment by assignUser** [comment]
This is ready now 🎉

**Comment by kgpai** [comment]
Importing to start with the build process internally..

**Comment by majetideepak** [comment]
@assignUser we must also upgrade fmt. Masha reported an issue in Prestissimo after upgrading folly but not fmt. I commented here https://github.com/facebookincubator/velox/issues/7896
I will respond to your question on the issue.

**Comment by kgpai** [comment]
@assignUser @majetideepak Is the fmt upgrade going to part of this diff or separate ?

**Comment by kgpai** [comment]
NVM, I see this pr : https://github.com/facebookincubator/velox/pull/7941

**Comment by assignUser** [comment]
Separate as it isn't just the version bump but needs additional changes #7941

**Comment by kgpai** [comment]
@majetideepak @assignUser Is this pr blocked on upgrading fmt ? I see no issues internally , but can hold of on landing it #7941 should go first.

**Comment by assignUser** [comment]
No this can be merged now, the fmt bump is necessary but unrelated outside of conda.

**Comment by majetideepak** [comment]
@kgpai Masha reported an issue in Prestissimo build when she only upgraded folly and not fmt.
It's reported here https://github.com/prestodb/presto/issues/21492#issuecomment-1843311216

**Comment by majetideepak** [comment]
Upgrading the fmt version fixed the build failure https://github.com/facebookincubator/velox/issues/7692#issuecomment-1845513934

**Comment by assignUser** [comment]
Ah, I forgot about that, sorry.

**Comment by assignUser** [comment]
I will rebase this once #7941 is merged

**Comment by assignUser** [comment]
@kgpai @majetideepak could one of you look at the test failures and benchmark segfault? These are out of my wheelhouse. This PR was green before the rebase (300+ new commits) on the fmt upgrade (but I tested it with fmt 10 before and it worked fine).

**Comment by kgpai** [comment]
@assignUser No worries, I can take a look.

**Comment by majetideepak** [comment]
@assignUser For some reason I see `folly::Init{&argc, &argv, false};` being used instead of `folly::Init init(&argc, &argv);`.
We discussed the problem here https://github.com/facebookincubator/velox/pull/7700#discussion_r1419248092

**Comment by assignUser** [comment]
Ah good catch I must have dropped the wrong commit during the rebase. I will fix in a bit, maybe that solves the problems already!

**Comment by majetideepak** [comment]
@assignUser If you can fix the `format-fix` job, this should be ready to be merged.

**Review by majetideepak — APPROVED** [review]

**Pull request closed without merging**
